File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        unordered_map<char, int> mpp;
        int maxi = 0, j = 0;

        for(int i=0;i<n;i++) {
            mpp[s[i]]++;

            while(mpp[s[i]] > 1) {
                mpp[s[j]]--;

                if(mpp[s[j]] == 0) {
                    mpp.erase(s[j]);
                }

                j++;
            }

            maxi = max(maxi, i - j + 1);
        }

        return maxi;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int len = 0, maxi = 0;
        Map<Character, Integer> freqMap = new HashMap<>();
        int j = 0;

        for (int i = 0; i < n; i++) {
            freqMap.put(s.charAt(i), freqMap.getOrDefault(s.charAt(i), 0) + 1);

            while (freqMap.get(s.charAt(i)) > 1) {
                freqMap.put(s.charAt(j), freqMap.getOrDefault(s.charAt(j), 0) - 1);

                if (freqMap.get(s.charAt(j)) == 0) {
                    freqMap.remove(s.charAt(j));
                }

                j++;
            }

            len = i - j + 1;
            maxi = Math.max(maxi, len);
        }

        return maxi;
    }
}